### ISSUE TITLE
[Phase 4] Streaming Infrastructure

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/high-dim-anomaly-detection#5]
## Summary

Implement streaming infrastructure with sliding windows and incremental updates. This phase enables real-time processing of continuous data streams.

**Part of:** Meta Issue #1
**Dependencies:** Phase 1 (data structures), Phase 2 (reduction), Phase 3 (detection)

## Objectives

1. Implement sliding window buffer management
2. Add Incremental PCA for streaming dimensionality reduction
3. Create stream processor orchestrating the pipeline
4. Ensure bounded memory usage

## Files to Create

| File | Lines (est.) | Action | Description |
|------|-------------|--------|-------------|
| `src/ingestion/sliding_window.py` | 100 | Create | Window management |
| `src/reduction/incremental_pca.py` | 80 | Create | Streaming PCA |
| `src/ingestion/stream_processor.py` | 150 | Create | Orchestration |
| `tests/test_streaming.py` | 120 | Create | Streaming tests |

### Sliding Window
- [ ] Implement `SlidingWindow` class with configurable size and stride
- [ ] Use `collections.deque` with maxlen for bounded buffer
- [ ] Track `is_ready()` state for complete windows
- [ ] Support `get_window()` returning numpy array
- [ ] Track total samples processed

### Incremental PCA
- [ ] Implement `IncrementalPCAReducer` wrapping sklearn IncrementalPCA
- [ ] Support `partial_fit()` for streaming updates
- [ ] Configure `n_components`, `batch_size` via YAML
- [ ] Track explained variance ratio

### Stream Processor
- [ ] Create `StreamProcessor` coordinating window -> reduce -> detect
- [ ] Accept data point by point or in batches
- [ ] Return anomaly scores and explanations per window
- [ ] Expose processing metrics (latency, throughput)

### Memory Management
- [ ] Enforce maximum buffer size
- [ ] Periodic garbage collection hints
- [ ] Profile memory under load

## Code Snippet: SlidingWindow

```python
# src/ingestion/sliding_window.py
"""Sliding window for streaming data."""

from collections import deque
from typing import Optional
import numpy as np


class SlidingWindow:
    """Fixed-size sliding window buffer."""

    def __init__(self, size: int, stride: int = 1, max_buffer: int = 10000):
        """
        Args:
            size: Samples per window
            stride: Samples between windows
            max_buffer: Maximum buffer size (memory bound)
        """
        if size <= 0:
            raise ValueError(f"Size must be positive, got {size}")
        if stride <= 0 or stride > size:
            raise ValueError(f"Stride must be in (0, size], got {stride}")

        self.size = size
        self.stride = stride
        self._buffer = deque(maxlen=max_buffer)
        self._samples_since_window = 0
        self._total_samples = 0

    def add(self, sample: np.ndarray) -> None:
        """Add single sample to buffer."""
        self._buffer.append(sample.copy())
        self._samples_since_window += 1
        self._total_samples += 1

    def add_batch(self, samples: np.ndarray) -> None:
        """Add multiple samples."""
        for sample in samples:
            self.add(sample)

    def is_ready(self) -> bool:
        """Check if window is complete and stride reached."""
        return (len(self._buffer) >= self.size and
                self._samples_since_window >= self.stride)

    def get_window(self) -> Optional[np.ndarray]:
        """Get current window (None if not ready)."""
        if not self.is_ready():
            return None

        window = np.array(list(self._buffer)[-self.size:])
        self._samples_since_window = 0
        return window

    def clear(self) -> None:
        """Reset buffer."""
        self._buffer.clear()
        self._samples_since_window = 0

    @property
    def buffer_size(self) -> int:
        return len(self._buffer)

    @property
    def total_samples_seen(self) -> int:
        return self._total_samples
```

## Code Snippet: IncrementalPCAReducer

```python
# src/reduction/incremental_pca.py
"""Incremental PCA for streaming reduction."""

import numpy as np
from sklearn.decomposition import IncrementalPCA
from .base_reducer import BaseReducer


class IncrementalPCAReducer(BaseReducer):
    """Streaming-compatible PCA using partial_fit."""

    def __init__(self, config: dict):
        super().__init__(config)
        ipca_config = config.get('reduction', {}).get('incremental', {})
        self.n_components = ipca_config.get('n_components', 15)
        self.batch_size = ipca_config.get('batch_size', 1000)

        self.model = IncrementalPCA(n_components=self.n_components)
        self._n_samples_seen = 0

    def fit(self, X: np.ndarray) -> 'IncrementalPCAReducer':
        """Full fit (for initial training)."""
        self.model.fit(X)
        self._n_samples_seen = len(X)
        self.is_fitted = True
        return self

    def partial_fit(self, X: np.ndarray) -> 'IncrementalPCAReducer':
        """Incremental update with new batch."""
        self.model.partial_fit(X)
        self._n_samples_seen += len(X)
        self.is_fitted = True
        return self

    def transform(self, X: np.ndarray) -> np.ndarray:
        if not self.is_fitted:
            raise RuntimeError("Reducer not fitted")
        return self.model.transform(X)

    def reconstruct(self, X: np.ndarray) -> np.ndarray:
        """Approximate reconstruction via inverse transform."""
        X_reduced = self.transform(X)
        return self.model.inverse_transform(X_reduced)

    @property
    def explained_variance_ratio(self) -> np.ndarray:
        return self.model.explained_variance_ratio_

    @property
    def n_samples_seen(self) -> int:
        return self._n_samples_seen
```

## Code Snippet: StreamProcessor

```python
# src/ingestion/stream_processor.py
"""Stream processor coordinating pipeline."""

from typing import Dict, List, Optional
from dataclasses import dataclass
import time
import numpy as np

from .sliding_window import SlidingWindow
from ..reduction.incremental_pca import IncrementalPCAReducer
from ..detection.ensemble import AnomalyEnsemble


@dataclass
class StreamResult:
    """Result from processing a window."""
    window_id: int
    timestamp: float
    scores: np.ndarray
    anomaly_count: int
    processing_time_ms: float


class StreamProcessor:
    """Process streaming data through anomaly detection pipeline."""

    def __init__(self, config: dict):
        self.config = config
        stream_config = config.get('streaming', {})

        self.window = SlidingWindow(
            size=stream_config.get('window_size', 1000),
            stride=stream_config.get('stride', 100)
        )

        self.reducer = IncrementalPCAReducer(config)
        self.ensemble = AnomalyEnsemble(config)
        self.threshold = config.get('detection', {}).get('ensemble', {}).get('threshold', 0.7)

        self._window_count = 0
        self._is_initialized = False

    def initialize(self, X_init: np.ndarray) -> None:
        """Initialize models with initial data."""
        self.reducer.fit(X_init)
        X_reduced = self.reducer.transform(X_init)
        self.ensemble.fit(X_reduced)
        self._is_initialized = True

    def process_sample(self, sample: np.ndarray) -> Optional[StreamResult]:
        """Process single sample, return result if window ready."""
        if not self._is_initialized:
            raise RuntimeError("Call initialize() first")

        self.window.add(sample)

        if not self.window.is_ready():
            return None

        return self._process_window()

    def process_batch(self, samples: np.ndarray) -> List[StreamResult]:
        """Process batch, returning all completed windows."""
        results = []
        for sample in samples:
            result = self.process_sample(sample)
            if result is not None:
                results.append(result)
        return results

    def _process_window(self) -> StreamResult:
        """Process a complete window."""
        start_time = time.time()

        window_data = self.window.get_window()
        self._window_count += 1

        # Update reducer incrementally
        self.reducer.partial_fit(window_data)

        # Reduce and score
        X_reduced = self.reducer.transform(window_data)
        scores = self.ensemble.score_samples(X_reduced)
        anomaly_count = (scores > self.threshold).sum()

        processing_time_ms = (time.time() - start_time) * 1000

        return StreamResult(
            window_id=self._window_count,
            timestamp=time.time(),
            scores=scores,
            anomaly_count=anomaly_count,
            processing_time_ms=processing_time_ms
        )
```

## Verification

```python
import numpy as np
from src.ingestion.stream_processor import StreamProcessor

# Configuration
config = {
    'streaming': {'window_size': 100, 'stride': 20},
    'reduction': {'incremental': {'n_components': 10}},
    'detection': {
        'isolation_forest': {'n_estimators': 50},
        'lof': {'n_neighbors': 10},
        'spc': {'regularization': 0.01},
        'ensemble': {'threshold': 0.7}
    }
}

# Generate stream
np.random.seed(42)
stream_data = np.random.randn(1000, 20)

# Initialize with first chunk
processor = StreamProcessor(config)
processor.initialize(stream_data[:200])

# Process stream
results = []
for sample in stream_data[200:]:
    result = processor.process_sample(sample)
    if result:
        results.append(result)
        print(f"Window {result.window_id}: {result.anomaly_count} anomalies, {result.processing_time_ms:.1f}ms")

print(f"Processed {len(results)} windows")
```

## Definition of Done

- [ ] Sliding window correctly manages data (size, stride)
- [ ] Incremental PCA updates without full refit
- [ ] Stream processor coordinates end-to-end
- [ ] Memory usage bounded regardless of stream length
- [ ] Latency < 100ms per window
- [ ] Unit tests pass

## Learning Notes

**Concepts to understand:**
- Sliding window algorithms
- Online/incremental learning
- Memory-bounded data structures (`deque` with `maxlen`)
- Latency vs throughput tradeoffs
